REGRESSION (291398@main): Slower String.prototype.split with result length < 100

ast-hugger opened the pull request:
#### d84f22e0d2e4d4139c884566b32262d459194e82
<pre>
REGRESSION (291398@main): Slower String.prototype.split with result length &lt; 100
<a href="https://bugs.webkit.org/show_bug.cgi?id=292147">https://bugs.webkit.org/show_bug.cgi?id=292147</a>
<a href="https://rdar.apple.com/150668072">rdar://150668072</a>

Reviewed by Yijia Huang and Yusuke Suzuki.

The performance regression was caused by two prior optimizations that involve adding JSStrings
to hash map in the VM, and the underlying AtomStringImpls to the atom string registry.
<a href="https://github.com/WebKit/WebKit/pull/37941">https://github.com/WebKit/WebKit/pull/37941</a>
<a href="https://github.com/WebKit/WebKit/pull/41448">https://github.com/WebKit/WebKit/pull/41448</a>

In the specific benchmark reported in the bug, both actions incur the cost of hashing and growing
these maps, which produces the observed regression. At the same time, the optimizations improve
performance in the targeted browser benchmark suites.

The change set introduces an additional condition for applying the optimizations. A string is only
promoted into an Identifier and registered with the VM if it can be a valid identifier (does not
begin with a digit). This preserves the optimization required for the common browser frameworks,
but avoids the costs in use case of parsing CSV files containing numeric data.

Additionally, there&apos;s a fix to defer GC for the duration of the registry update in the empty string
code path.

* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: <a href="https://commits.webkit.org/294689@main">https://commits.webkit.org/294689@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a93c55bb0687125ed079e55f5a5ded3d73ea0766

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102667 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22340 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12658 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107832 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53308 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22648 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30842 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78091 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35068 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105673 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17548 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92654 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58423 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17390 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10687 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52665 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/95342 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87209 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10755 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110208 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/101277 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29804 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21978 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87075 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30168 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88845 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86680 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22073 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31506 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9232 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/24057 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29731 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/35051 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/124910 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29539 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34673 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32865 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31101 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/runtime/StringPrototype.cpp:
```diff
@@ -706,7 +706,7 @@ JSC_DEFINE_HOST_FUNCTION(stringProtoFuncSplitFast, (JSGlobalObject* globalObject
         unsigned resultSize = result.size();
         if (limit == 0xFFFFFFFFu && !globalObject->isHavingABadTime() && resultSize < MIN_SPARSE_ARRAY_INDEX) [[likely]] {
-            bool makeAtomStringsArray = resultSize < atomStringsArrayLimit;
+            bool makeAtomStringsArray = false; // temporarily disabled because of a performance regression, see PR comment for details
```
We should make this comment: 
```
// FIXME: Temporarily disabled because of a performance regression. See: <link to bugzilla to fix it with explanation of what went wrong>
```

That way folks don't have to dig through the commit history. It also makes it harder to forget to file a bug to actually fix this :)

Constellation commented on Source/JavaScriptCore/runtime/StringPrototype.cpp:
```diff
@@ -706,7 +706,7 @@ JSC_DEFINE_HOST_FUNCTION(stringProtoFuncSplitFast, (JSGlobalObject* globalObject
         unsigned resultSize = result.size();
         if (limit == 0xFFFFFFFFu && !globalObject->isHavingABadTime() && resultSize < MIN_SPARSE_ARRAY_INDEX) [[likely]] {
-            bool makeAtomStringsArray = resultSize < atomStringsArrayLimit;
+            bool makeAtomStringsArray = false; // temporarily disabled because of a performance regression, see https://bugs.webkit.org/show_bug.cgi?id=292147
```
Making registered strings AtomString was done before 291398@main. 291398@main was adding these registered JSStrings to the VM's table (and getting some strings from that table). So flipping this flag to false causes significant issue which disables the things which is not introduced from 291398@main

Constellation reviewed (changes requested):
So, as the bugzilla report says, we should disable the part which is registering JSString to WeakGCMap table in VM.

Constellation commented on Source/JavaScriptCore/runtime/StringPrototype.cpp:
```diff
@@ -722,7 +722,7 @@ JSC_DEFINE_HOST_FUNCTION(stringProtoFuncSplitFast, (JSGlobalObject* globalObject
                 unsigned end = result[i];
                 JSString* string = nullptr;
-                if (makeAtomStringsArray) {
+                if (makeAtomStringsArray && !u_isdigit(view->characterAt(start))) {
```
Let's use `isASCIIDigit`

hyjorc1 reviewed (approved):
r=me

Constellation commented on Source/JavaScriptCore/runtime/StringPrototype.cpp:
```diff
@@ -722,7 +722,7 @@ JSC_DEFINE_HOST_FUNCTION(stringProtoFuncSplitFast, (JSGlobalObject* globalObject
                 unsigned end = result[i];
                 JSString* string = nullptr;
-                if (makeAtomStringsArray) {
+                if (makeAtomStringsArray && !isASCIIDigit(view->characterAt(start))) {
```
When doing this, we are relying on the fact that JSImmutableButterfly with a specific Structure always have atomized string. Thus I think we need some structure change when we hit this case.

Jarred-Sumner commented on Source/JavaScriptCore/runtime/StringPrototype.cpp:
```diff
@@ -719,10 +719,11 @@ JSC_DEFINE_HOST_FUNCTION(stringProtoFuncSplitFast, (JSGlobalObject* globalObject
                 unsigned end = result[i];
                 JSString* string = nullptr;
-                if (makeAtomStringsArray) {
+                if (makeAtomStringsArray && !(start < end && isASCIIDigit(view->characterAt(start)))) {
```
It would be more effective to check if it’s not an identifier start, which for ascii is $A-Za-z_. So anything not in that range which also excludes characters like ^ and importantly quote characters like `”` `’` and commas

ast-hugger replied:
Good idea, done.

ast-hugger committed:
REGRESSION (291398@main): Slower String.prototype.split with result length < 100
https://bugs.webkit.org/show_bug.cgi?id=292147
rdar://150668072

Reviewed by Yijia Huang and Yusuke Suzuki.

The performance regression was caused by two prior optimizations that involve adding JSStrings
to hash map in the VM, and the underlying AtomStringImpls to the atom string registry.
https://github.com/WebKit/WebKit/pull/37941
https://github.com/WebKit/WebKit/pull/41448

In the specific benchmark reported in the bug, both actions incur the cost of hashing and growing
these maps, which produces the observed regression. At the same time, the optimizations improve
performance in the targeted browser benchmark suites.

The change set introduces an additional condition for applying the optimizations. A string is only
promoted into an Identifier and registered with the VM if it can be a valid identifier (does not
begin with a digit). This preserves the optimization required for the common browser frameworks,
but avoids the costs in use case of parsing CSV files containing numeric data.

Additionally, there's a fix to defer GC for the duration of the registry update in the empty string
code path.

* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: https://commits.webkit.org/294689@main
